r/src/array_from_vector.cpp

```diff
@@ -351,15 +351,20 @@ std::shared_ptr<Array> MakeFactorArrayImpl(Rcpp::IntegerVector_ factor,
 
 std::shared_ptr<Array> MakeFactorArray(Rcpp::IntegerVector_ factor,
                                        const std::shared_ptr<arrow::DataType>& type) {
-  SEXP levels = factor.attr("levels");
-  int n = Rf_length(levels);
-  if (n < 128) {
+  auto* dict_type = checked_cast<arrow::DictionaryType*>(type.get());
```

Comment:
Small nit, we tend to prefer `const &` instead of pointer. This indicates to the reader that you won't mutate the object.

```suggestion
  auto dict_type = checked_cast<const arrow::DictionaryType&>(*type);
```

Reply:
That would be `const auto& dict_type` then. Otherwise +1 to what @fsaintjacques said.